pkg/cluster/internal/create/kubeadm-join.go

```diff
@@ -40,49 +43,130 @@ func newKubeadmJoinAction() Action {
+	}
+
+	// get the target node for this task (the joining node)
+	node, ok := ec.NodeFor(configNode)
 	if !ok {
-		return fmt.Errorf("unable to get the handle for operating on node: %s", ec.DerivedConfig.BootStrapControlPlane().Name)
+		return fmt.Errorf("unable to get the handle for operating on node: %s", configNode.Name)
 	}
 
-	// gets the IP of the bootstrap master node
-	controlPlaneIP, err := controlPlaneHandle.IP()
+	// creates the folder tree for pre-loading necessary cluster certificates
+	// on the joining node
+	if err := node.Command("mkdir", "-p", "/etc/kubernetes/pki/etcd").Run(); err != nil {
+		return errors.Wrap(err, "failed to join node with kubeadm")
+	}
+
+	// define the list of necessary cluster certificates
+	fileNames := []string{
+		"ca.crt", "ca.key",
+		"front-proxy-ca.crt", "front-proxy-ca.key",
+		"sa.pub", "sa.key",
+	}
+	if ec.ExternalEtcd() == nil {
+		fileNames = append(fileNames, "etcd/ca.crt", "etcd/ca.key")
+	}
+
+	// creates a temporary folder on the host that should acts as a transit area
+	// for moving necessary cluster certificates
+	tmpDir, err := fs.TempDir("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	err = os.MkdirAll(filepath.Join(tmpDir, "/etcd"), os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// get the handle for the bootstrap control plane node (the source for necessary cluster certificates)
+	controlPlaneHandle, ok := ec.NodeFor(ec.BootStrapControlPlane())
+	if !ok {
+		return fmt.Errorf("unable to get the handle for operating on node: %s", ec.BootStrapControlPlane().Name)
+	}
+
+	// copies certificates from the bootstrap control plane node to the joining node
+	for _, fileName := range fileNames {
+		// sets the path of the certificate into a node
+		containerPath := filepath.Join("/etc/kubernetes/pki", fileName)
+		// set the path of the certificate into the transit are on the host
+		tmpPath := filepath.Join(tmpDir, fileName)
+		// copies from bootstrap control plane node to tmp area
+		if err := controlPlaneHandle.CopyFrom(containerPath, tmpPath); err != nil {
+			return errors.Wrapf(err, "failed to copy certificate %s", fileName)
+		}
+		// copies from tmp area to joining node
+		if err := node.CopyTo(tmpPath, containerPath); err != nil {
+			return errors.Wrapf(err, "failed to copy certificate %s", fileName)
+		}
+	}
+
+	// run kubeadm join --control-plane
+	if err := node.Command(
+		"kubeadm", "join",
+		// the join command uses the docker ip and a well know port that
+		// are accessible only inside the docker network
+		joinAddress,
+		// set the node to join as control-plane
+		"--experimental-control-plane",
+		// uses a well known token and skips ca certification for automating TLS bootstrap process
+		"--token", kubeadm.Token,
+		"--discovery-token-unsafe-skip-ca-verification",
+		// preflight errors are expected, in particular for swap being enabled
+		// TODO(bentheelder): limit the set of acceptable errors
+		"--ignore-preflight-errors=all",
```

Comment:
possibly, OTOH I'd rather continue on for now and handle actual failures. I've been meaning to write some of these things down in the dev guide, graceful degradation is a core tenent so far.

environmental checks are also generally expected to fail given our environment not being technically supported for many reasons.

I consider this similar to how we do not fail on ensuring the images and instead fail if `docker run $img` fails - kind will continue on until it can't :^)